src/duckdb.mjs

```diff
@@ -1,361 +1,228 @@
+      throw error;
     }
-
-    async queryStream(query, params) {
-      const conn = await this.connection();
-      let result;
-
-      if (params) {
-        const stmt = await conn.prepare(query);
-        result = await stmt.query(...params);
-      } else {
-        result = await conn.query(query);
-      }
-      // Populate the schema of the results
-      const schema = result.schema.fields.map(({
-        name,
-        type
-      }) => ({
+    return {
+      schema: schema.fields.map(({name, type}) => ({
         name,
         type: getType(String(type)),
         databaseType: String(type)
-      }));
-      return {
-        schema,
-        async *readRows() {
-          let rows = result.toArray().map((r) => Object.fromEntries(r));
-          yield rows;
-        }
-      };
-    }
-
-    // This function gets called to prepare the `query` parameter of the `queryStream` method
-    queryTag(strings, ...params) {
-      return [strings.join("?"), params];
-    }
-
-    escape(name) {
-      return `"${name}"`;
-    }
-
-    async describeTables() {
-      const conn = await this.connection();
-      const tables = (await conn.query(`SHOW TABLES`)).toArray();
-      return tables.map(({
-        name
-      }) => ({
-        name
-      }));
-    }
-
-    async describeColumns({
-      table
-    } = {}) {
-      const conn = await this.connection();
-      const columns = (await conn.query(`DESCRIBE ${table}`)).toArray();
-      return columns.map(({
-        column_name,
-        column_type
-      }) => {
-        return {
-          name: column_name,
-          type: getType(column_type),
-          databaseType: column_type
-        };
-      });
-    }
-
-    async db() {
-      if (!this._db) {
-        this._db = await makeDB();
-        await this._db.open({
-          query: {
-            castTimestampToDate: true
+      })),
+      async *readRows() {
+        try {
+          while (!batch.done) {
+            yield batch.value.toArray();
+            batch = await reader.next();
           }
-        });
-      }
-      return this._db;
-    }
-
-    async connection() {
-      if (!this._conn) {
-        const db = await this.db();
-        this._conn = await db.connect();
-      }
-      return this._conn;
-    }
-
-    async reconnect() {
-      if (this._conn) {
-        this._conn.close();
+        } finally {
+          await connection.close();
+        }
       }
-      delete this._conn;
-    }
-
-    // The `.queryStream` function will supercede this for SQL and Table cells
-    // Keeping this for backwards compatibility
-    async query(query, params) {
-      const conn = await this.connection();
-      let result;
+    };
+  }
 
-      if (params) {
-        const stmt = await conn.prepare(query);
-        result = stmt.query(...params);
-      } else {
-        result = await conn.query(query);
+  async query(query, params) {
+    const result = await this.queryStream(query, params);
```

Comment:
Do we need to have an implementation of `.query` that doesn't rely on `queryStream` for some back-compatibility with existing DuckDB notebooks (e.g., [this cel](https://observablehq.com/@cmudig/duckdb-client#cell-16)l)? I imagine it's fine, as those imported versions of `DuckDBClient` will overwrite the standard library version, but just wanted to surface this as a possible concern.

Reply:
(leaving some notes as we're pairing on the review, thinking out loud a bit) - does this mean that all DuckDBClients support streaming (as this uses the `queryStream` method)?

Reply:
Anything in the standard library can be overridden, so any notebook that defines its own DuckDBClient won’t see the implementation here.

As far as this implementation goes, the fact that it implements _db_.query on top of _db_.queryStream should be totally invisible to the user. It implements the _db_.query method as specified in the database client specification and it only calls _db_.queryStream under the hood to reduce code duplication. It does not somehow mean that _db_.query or _db_.queryRow supports streaming; those methods still just return a promise to the results.

I should also note that I didn’t implement the signal option for query methods in the DuckDBClient so there isn’t currently a way to cancel queries. But in practice that probably shouldn’t matter much because it’s an in-memory client and therefore most queries should be relatively fast. We could add it in the future.